pkg/controllers/management/auth/user.go

```diff
@@ -0,0 +1,235 @@
+
+import (
+	"fmt"
+
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+type userLifecycle struct {
+	projectRoles        v3.ProjectRoleTemplateBindingInterface
+	clusterRoles        v3.ClusterRoleTemplateBindingInterface
+	globalRoles         v3.GlobalRoleBindingInterface
+	projectRolesLister  v3.ProjectRoleTemplateBindingLister
+	clusterRolesLister  v3.ClusterRoleTemplateBindingLister
+	globalRolesLister   v3.GlobalRoleBindingLister
+	projectRolesIndexer cache.Indexer
+	clusterRolesIndexer cache.Indexer
+	globalRolesIndexer  cache.Indexer
+}
+
+const clusterRoleBindingByUserRefKey = "auth.management.cattle.io/crb-by-user-ref"
+const projectRoleBindingByUserRefKey = "auth.management.cattle.io/prb-by-user-ref"
+const globalRoleBindingByUserRefKey = "auth.management.cattle.io/grb-by-user-ref"
+
+func newUserLifecycle(management *config.ManagementContext) *userLifecycle {
+	lfc := &userLifecycle{
+		projectRoles:       management.Management.ProjectRoleTemplateBindings(""),
+		clusterRoles:       management.Management.ClusterRoleTemplateBindings(""),
+		globalRoles:        management.Management.GlobalRoleBindings(""),
+		projectRolesLister: management.Management.ProjectRoleTemplateBindings("").Controller().Lister(),
+		clusterRolesLister: management.Management.ClusterRoleTemplateBindings("").Controller().Lister(),
+		globalRolesLister:  management.Management.GlobalRoleBindings("").Controller().Lister(),
+	}
+
+	projectRoleBindingInformer := management.Management.ProjectRoleTemplateBindings("").Controller().Informer()
+	projectRoleBindingInformer.AddIndexers(map[string]cache.IndexFunc{
+		projectRoleBindingByUserRefKey: projectRoleBindingByUserRefFunc,
+	})
+
+	lfc.projectRolesIndexer = projectRoleBindingInformer.GetIndexer()
+
+	clusterRoleBindingInformer := management.Management.ClusterRoleTemplateBindings("").Controller().Informer()
+	clusterRoleBindingInformer.AddIndexers(map[string]cache.IndexFunc{
+		clusterRoleBindingByUserRefKey: clusterRoleBindingByUserRefFunc,
+	})
+
+	lfc.clusterRolesIndexer = clusterRoleBindingInformer.GetIndexer()
+
+	globalRoleBindingInformer := management.Management.GlobalRoleBindings("").Controller().Informer()
+	globalRoleBindingInformer.AddIndexers(map[string]cache.IndexFunc{
+		globalRoleBindingByUserRefKey: globalRoleBindingByUserRefFunc,
+	})
+
+	lfc.globalRolesIndexer = globalRoleBindingInformer.GetIndexer()
+
+	return lfc
+}
+
+func globalRoleBindingByUserRefFunc(obj interface{}) ([]string, error) {
+	globalRoleBinding, ok := obj.(*v3.GlobalRoleBinding)
+	if !ok {
+		return []string{}, nil
+	}
+
+	return []string{globalRoleBinding.UserName}, nil
+}
+
+func projectRoleBindingByUserRefFunc(obj interface{}) ([]string, error) {
+	projectRoleBinding, ok := obj.(*v3.ProjectRoleTemplateBinding)
+	if !ok {
+		return []string{}, nil
+	}
+
+	return []string{projectRoleBinding.UserName}, nil
```

Comment:
There is a scenario where projectRoleTemplateBinding and clusterRoleTemplateBinding can have an _unset_ UserName field. So, i think these indexes should return an empty slice in that scenario, not a slice of length 1 with a value of "".

Note that this does not apply to globalRoleBinding. it willl always have userName set